module/modmanager/manager.go

```diff
@@ -0,0 +1,370 @@
+// Package modmanager provides the module manager for a robot.
+package modmanager
+
+import (
+	"context"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/edaniels/golog"
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+	pb "go.viam.com/api/module/v1"
+	"go.viam.com/utils/pexec"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"go.viam.com/rdk/config"
+	rdkgrpc "go.viam.com/rdk/grpc"
+	modlib "go.viam.com/rdk/module"
+	modif "go.viam.com/rdk/module/modmanager/modmaninterface"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/robot"
+)
+
+// NewManager returns a Manager.
+func NewManager(r robot.LocalRobot) (modif.ModuleManager, error) {
+	mgr := &Manager{
+		mu:      sync.RWMutex{},
+		logger:  r.Logger(),
+		modules: make(map[string]*module),
+		r:       r,
+		rMap:    make(map[resource.Name]*module),
+	}
+	return mgr, nil
+}
+
+type module struct {
+	name    string
+	process pexec.ManagedProcess
+	handles modlib.HandlerMap
+	conn    *grpc.ClientConn
+	client  pb.ModuleServiceClient
+	addr    string
+}
+
+// Manager is the root structure for the module system.
+type Manager struct {
+	mu      sync.RWMutex
+	logger  golog.Logger
+	modules map[string]*module
+	r       robot.LocalRobot
+	rMap    map[resource.Name]*module
+}
+
+// Close terminates module connections and processes.
+func (mgr *Manager) Close(ctx context.Context) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	var err error
+	for _, mod := range mgr.modules {
+		err = multierr.Combine(
+			err,
+			mod.conn.Close(),
+			err, mod.process.Stop(),
+		)
+	}
+	return err
+}
+
+// AddModule adds and starts a new resource module.
+func (mgr *Manager) AddModule(ctx context.Context, cfg config.Module) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	_, exists := mgr.modules[cfg.Name]
+	if exists {
+		return nil
+	}
+
+	mod := &module{}
+	mgr.modules[cfg.Name] = mod
+
+	parentAddr, err := mgr.r.LiteAddress()
+	if err != nil {
+		return err
+	}
+	mod.addr = filepath.Dir(parentAddr) + "/" + cfg.Name + ".sock"
+
+	pcfg := pexec.ProcessConfig{
+		ID:   cfg.Name,
+		Name: cfg.Path,
+		Args: []string{mgr.modules[cfg.Name].addr},
+		Log:  true,
+	}
+	mod.process = pexec.NewManagedProcess(pcfg, mgr.logger)
```

Comment:
similar but better placed comment about" what testing/assurances do we have when a process dies?

Reply:
I'm making a TODO note about this topic, but I believe this fits in the next PR scope better. As mentioned in a previous comment, once "reconfigure" of a module works, reloading config into a crashed/restarted module should be trivial, and this can be shored up.

For now, this breaks pretty bad if the module itself dies. Basically a restart (of the parent) is required. But the same is true if the parent process dies, so from a practical standpoint, crashes that kill the process in a module are the same as crashing code in the parent.

Reply:
sounds good